### PR TITLE
⚙️ Add smart tracking config

### PR DESCRIPTION
**Pull request by shantanuraj**
## :star2: What does this PR do?

- 📦 Add base config fixture
- 🔧 Add config command
- 🔍 Add config lookup
- 👩‍🎨 Add config init command
- 📐 Refactor utilities
- 🩹 Fix description help not showing up

## :bug: Recommendations for testing

- See recording for quick demo.
  [asciicast](https://asciinema.org/a/ZkXHPgB7qAs5spFTLPghjXPl0)
- Another recording shows how configuration is inherited by all child directories but not the ancestors of a tracked dir.
  [asciicast](https://asciinema.org/a/UlqtsLDukPWfkVSDc6ks4qloS)
- Adds `toggl config` command to manage auto tracking configuration [^1].
- The focus of this PR is to implement a neat way to manage your auto-tracking configuration toml, and not the configuration itself.
- We also add a few fixtures to test the config logic at some point.
- The default configuration[^2] is bundled into the binary.

### New commands introduced

```console
# Create a new configuration using the default bundled config
$ toggl config init

# Pretty print currently active configuration
$ toggl config

# Edit the configuration in the system's default editor
$ toggl config --edit # or toggl config -e

# Delete the active configuration for current directory
$ toggl config --delete # or toggl config -d

# Log the path of the active configuration file
$ toggl config --path # or toggl config -p
```

## :memo: Links to relevant issues/docs

See #45 which defines the configuration storage and retrieval mechanism.
Minor changes have been incorporated that differ from the original spec.

Instead of using `md5(path)` we use `base64(path)`
  - because md5 usage is highly discouraged by the ecosystem.
  - we don't need the file path to be hashed just encoded.
  - another benefit is, that the encoded is reversible, so if we plan to add something like `toggl config list` to log all tracked paths; it would be trivial.
Config files are stored in directories specified by the `directories::ProjectDirs::from("studio.watercooler", "labs", "toggl-cli")`[^3] method.

I've updated the spec to reflect these changes.

Current locations are

| linux                                                     | windows                                           | macOS                                                                 |
| --------------------------------------------------------- | ------------------------------------------------- | --------------------------------------------------------------------- |
| `$XDG_CONFIG_HOME`/toggl-cli or `$HOME`/.config/toggl-cli | `{FOLDERID_RoamingAppData}`/labs/toggl-cli/config | `$HOME`/Library/Application Support/studio.watercooler.labs.toggl-cli |

## :speech_balloon: Summarise how you feel about this PR with a gif/image

ohstunning-sbagliato

[^1]: Auto tracking RFC at #45
[^2]: [default.toml](./src/config/fixtures/default.toml)
[^3]: For platform agnostic config directory calculation https://docs.rs/directories/5.0.0/directories/struct.ProjectDirs.html#method.config_dir

### CODE REVIEW

File: src/config/fixtures/default.toml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# https://support.toggl.com/en/articles/2452474-introduction-to-workspaces#switching-and-creating-workspaces
+# workspace = "My Workspace"
+
+# Description (optional, default="")
+# Accepts any string template with our marcros in them
+description = "🔨 {{current_branch}}"
+
+# Project (optional, default=No project)
+# Accepts any string template with our marcros in them
+# A project is created if it doesn't exist
```
Comment (tcrammond) [inline src/config/fixtures/default.toml:21]: That may fail if the user might not have permission in the workspace, wil just need to handle the rejected request: https://support.toggl.com/en/articles/2216605-access-rights-and-privileges#:~:text=A%20regular%20user%20can%3A&text=Add%20projects%20and%20clients%20(unless,Add%20tags. Or could also support a mapping/remember the project last assigned to a directory (similar to official browser extension)

**Comment by tcrammond** [comment]
Just chiming in that it's functioning beautifully for me 💅

File: src/config/fixtures/default.toml (hunk 1 above)
Comment (shantanuraj) [reply src/config/fixtures/default.toml:21]: Good point, will track for when we implement the last bit that actually ensures all entities are present and creates the time-entry from the config.

Comment (shantanuraj) [reply src/config/fixtures/default.toml:21]: > Or could also support a mapping/remember the project last assigned to a directory (similar to official browser extension

That would be achieved it they manually add the project name (of one that already exists)
```toml
project = "toggl-cli"
```

or are you suggestion another template variable they could utilize instead?

Comment (heytherewill) [reply src/config/fixtures/default.toml:21]: If we're not creating the entities yet we should remove the comment from the default toml

File: src/config/init.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                }
+            }
+            Err(_) => {
+                let default_config = include_bytes!("./fixtures/default.toml");
+                let config_path = super::locate::get_config_path_for_current_dir()?;
+                let config_dir = config_path.parent().unwrap();
+                let msg = format!(
+                    "{} {}",
+                    "Created config at".green().bold(),
+                    config_path.display()
```
Comment (heytherewill) [inline src/config/init.rs:30]: Maybe shorten this to `~/.toggl-cli/path` on Unix and `%localappdata%/.toggl-cli/path` on windows

File: src/config/manage.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+use colored::Colorize;
+
+use crate::{models::ResultWithDefaultError, utilities};
+
+pub struct ConfigManageCommand;
+
+impl ConfigManageCommand {
+    pub async fn execute(delete: bool, edit: bool, show_path: bool) -> ResultWithDefaultError<()> {
+        match (delete, edit, show_path) {
```
Comment (heytherewill) [inline src/config/manage.rs:11]: Use ifs here for better readability

File: src/config/manage.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                let path = super::locate::locate_config_path()?;
+                println!("{}", path.display());
+                Ok(())
+            }
+            _ => {
+                match super::locate::locate_config() {
+                    Ok(config) => {
+                        println!("{}", config);
+                    }
+                    Err(_) => {
```
Comment (heytherewill) [inline src/config/manage.rs:37]: We need to parse the error here and show that a file actually exists and is invalid rather than saying it doesn't exist

**Review by heytherewill — CHANGES_REQUESTED** [review]

File: src/config/fixtures/default.toml (hunk 1 above)
Comment (shantanuraj) [reply src/config/fixtures/default.toml]: Dropped the comment.

**Review by heytherewill — DISMISSED** [review]
Nice <3

**Review by heytherewill — APPROVED** [review]
